### ISSUE TITLE
Name field for non-input components

### ISSUE THREAD

**Issue by mackorone** [jppellet/Logic-Circuit-Simulator#15]
Hi! First and foremost, thanks for creating and maintaining this project. I'm currently using it to audit my 1Password account security. For example, both the "secret key" and "master password" are required for account access, so "account access" can be modeled as an AND gate with those two inputs.

Anyway, I'm trying to figure out how to label my components so that:
1. The diagram is readable in the browser
2. The downloaded JSON file can be processed by a program

The second item is the one I'm struggling with. In particular, only the input nodes have a `name` field that gets displayed alongside the component. The other nodes do have an `id` field, but it's not displayed. I'm currently using labels to give names to non-input components, but it's difficult for my program to correctly determine which labels correspond to which components from the JSON file alone. I could *also* store the name in the `id` field for the component, but that's redundant and easy to forget.

Would you consider adding a `name` field to *all* components, for the sake of simplifying this use case?

**Comment by jppellet** [comment]
Hey Mack, good idea. Will put that on the list. At some point, I had the idea, to ease layout, to be able to anchor some component B to another component A, so that when you move A, B is moved as well. What do you think: is it easier to just add a name field to all components or to allow a label to be anchored to a component (and thus reference that component in the JSON)? It would be a little bit more difficult to parse, but it would allow free placement of the label B for component A instead of a precomputed position (and also allow changing fonts, etc.). Computing the position automatically can be challenging as some components have inputs/outputs on all four sides.

**Comment by mackorone** [comment]
@jppellet Either approach (or both) works for me!

To give a bit more context, my program works with YAML files that look like this:
```
account_access:
- - authorized_device
  - master_password

authorized_device:
- - computer
- - phone
```
Corresponding to:
```
account_access = authorized_device AND master_password
authorized_device = computer OR phone
```

I'm currently working on the logic to convert back-and-forth between my YAML format and your JSON format. I need to compute ideal positions for each of the nodes in the circuit and then encode those positions in the JSON. *Also* computing the ideal position of each label shouldn't be too much additional work, and may actually provide me more flexibility regarding the layout. Plus, it's probably easier for me to solve the label-placement problem for just the components that I use - IN, OUT, AND, OR, BUFFER - than for you to add generic support for all components.

That said, if your program just-so-happened to automatically layout the labels *and* nodes in an ideal way, it would save me some work. Plus, then it wouldn't require a round trip to my YAML format and back to organize the circuit.

So in summary, my preferences are ranked as follows:
1. An easy and unambiguous way to label each node in the circuit (both solutions accomplish this)
1. The ability to manually specify the locations of nodes and labels (so I can override if necessary)
1. Automatic ideal label placement, relative to each node (my original question, may save me some work)
1. Automatic ideal node placement, relative to other nodes (and extension of my question, may save me some work)

**Comment by jppellet** [comment]
Alright, good points, I'll keep it in my TODO list and will come back to this issue. It is unfortunately unlikely that I will have much time for this in the coming weeks, though.

**Comment by jppellet** [comment]
OK, I have added the anchoring feature, finally, in 4674275531721a7fbd15cdeb11611b91d8c524d3.

So basically, components can now have a field called 'anchor' which references the ID of another component. You can now more easily parse the label of a node by checking if there is a label whose 'anchor' is the id of the component.

For instance here, "label" is the label for the input with id `in0`:

    { // JSON5
      v: 6,
      components: {
        in0: {type: 'in', pos: [140, 155], id: 17},
        out0: {type: 'out', pos: [365, 195], id: 18},
        label0: {type: 'label', pos: [145, 100], anchor: 'in0', text: 'Label'},
      },
      wires: [[17, 18]]
    }

In the user interface, the label will automatically move when its anchor is moved.

To set anchors in the UI, Command-drag from the label to its anchor.

Hope this helps, even if late!

**Comment by mackorone** [comment]
@jppellet this is awesome, thanks for implementing it!

> To set anchors in the UI, Command-drag from the label to its anchor.

This is going to sound silly, but I don't have a "command" button on my keyboard. Would it be possible to add alternative controls, like Control-drag or Shift-drag, or even just a way to add an anchor via the right-click menu?

**Comment by jppellet** [comment]
Ah, sure, it's not actually exactly the Command key, it's what the code sees as the Meta key. I suppose it is the Windows key on Windows—assuming you are on Windows?…. It can't be Control because it triggers the context menu, and Shift is a move that is constrained to be along the vertical axis or horizontal axis only. Does any modifier key currently work for you?

**Comment by mackorone** [comment]
@jppellet I currently only have the Command and Shift modifier keys available. (I'm on Ubuntu 22.04, and I remap everything to one of those two.) Would the right-click menu be doable? If not, no worries, just figured I'd toss the idea out there.

**Comment by jppellet** [comment]
I can't really make dragging from a menu item from the context menu possible easily without a bunch of new code, unfortunately…

**Comment by mackorone** [comment]
@jppellet no worries, I can figure something out on my end. Thanks again for implementing this!

**Comment by jppellet** [comment]
Alright!

**Comment by jppellet** [comment]
So, I've found an OK way to add an element to the context menu now (1539d0a4219a91f49ee71667c3cd69c0c63a8564). You can now choose Set Anchor and click on a component.

**Comment by mackorone** [comment]
@jppellet I just tested it out, and it works perfectly! Thank you so much!
